grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQueryIterator.java

```diff
@@ -90,30 +99,56 @@
         if (ruleIterator.hasNext()) subGoals.add(query);
     }
 
+    private Iterator<Answer> getRuleQueryIterator(InferenceRule rule, Unifier u, Unifier pu){
+        LOG.trace("Applying rule to: " + query +
+                rule + "\n" +
+                "t = " + u + "\n" +
+                "tp = " + pu);
+
+        //delta' = theta . thetaP . delta
+        Unifier uInv = u.inverse();
+        Answer partialSubPrime = query.getSubstitution()
+                .unify(pu)
+                .unify(uInv);
+
+        Set<VarName> headVars = rule.getHead().getVarNames();
+        Iterator<Answer> baseIterator = rule.getBody().iterator(partialSubPrime, subGoals, cache);
+        return Iterators.transform(
+                baseIterator,
+                a -> {
+                    if (a == null) return null;
+                    else {
+                        return a
```

Comment:
What does this actually do? I have been staring at it for quite some time and can't figure it out. Maybe a comment stating what this does?